File: backend/app/auth/features.py
"""Canonical feature list and role-based defaults for per-user feature access."""

import json
from typing import Optional

FEATURE_CATEGORIES: dict[str, list[str]] = {
    "CRM": ["contacts", "pipeline", "tasks", "cards"],
    "Sales": ["invoices", "estimates", "proposals"],
    "Accounting": [
        "cashbook",
        "expenses",
        "smart_import",
        "email_scanner",
        "reports",
        "tax",
        "recurring",
    ],
    "Communication": ["inbox", "phone", "sms"],
    "Automation": ["workflows", "forms"],
    "Content": ["pages", "docs", "sheets", "slides"],
    "Storage": ["drive"],
    "Meetings": ["calendar", "meeting_rooms"],
    "AI": ["obrain_chat", "obrain_coach"],
    "Admin": ["platform_admin", "portal_admin"],
}
# `estimates`, `expenses`, `tasks`, `workflows` and `forms` were added when the
# module gate was actually enforced: those routers previously had NO feature key,
# so there was no way to switch them off for an employee. Adding a key here
# automatically widens the admin checkbox grid — the UI reads this dict.

ALL_FEATURES: list[str] = [f for feats in FEATURE_CATEGORIES.values() for f in feats]

_ALL_TRUE = {f: True for f in ALL_FEATURES}
_ALL_FALSE = {f: False for f in ALL_FEATURES}

ROLE_DEFAULTS: dict[str, dict[str, bool]] = {
    "admin": {**_ALL_TRUE},
    # Same modules as a team member. A manager's extra power is VISIBILITY (it
    # sees its direct reports' records), not extra sections.
    "manager": {**_ALL_TRUE, "platform_admin": False},
    "team_member": {**_ALL_TRUE, "platform_admin": False},
    # An accountant runs the money side end to end: they handle the invoices that
    # come in from email, so they get invoicing (invoices + estimates) and the
    # contacts to invoice against, on top of the accounting modules. contacts is
    # only the MODULE — records stay owner-private, so they still see only their
    # own contacts, not the whole agency's.
    "accountant": {
        **_ALL_FALSE,
        "cards": True,
        "cashbook": True,
        "expenses": True,
        "invoices": True,
        "estimates": True,
        "contacts": True,
        "smart_import": True,
        "email_scanner": True,
        "reports": True,
        "tax": True,
        "recurring": True,
        "drive": True,
    },
    "client": {**_ALL_FALSE},
    # A viewer is a read-only collaborator: it owns nothing and sees only what is
    # explicitly SHARED with it. It still needs the contacts module switched on,
    # or the module gate would block the very share it was granted — it would be
    # handed a record it cannot open.
    "viewer": {**_ALL_FALSE, "contacts": True},
}

#: These are DEFAULTS, not policy. An admin overrides any of them per user via
#: feature_access_json (PlatformAdminPage → FeatureAccessEditor).


def resolve_feature_access(
    role: str,
    feature_access_json: Optional[str] = None,
) -> dict[str, bool]:
    """Resolve effective features.

    No explicit selection  -> the role's default preset.
    An explicit selection  -> AUTHORITATIVE and FAIL-CLOSED: a feature the
                              selection does not mention is NOT granted.

    The fail-closed half matters. This previously merged the selection *over* the
    role defaults, so any key the selection omitted kept its default — and
    team_member/manager default to every feature true. Consequences seen in
    practice:
      * selecting a single module for a team_member granted 29 features, and the
        account could open modules the admin never picked;
      * a stale cached frontend bundle submitting a grid from before a feature key
        existed silently granted that new feature to everyone it created.
    An admin who wants a role's full preset simply leaves the selection unset.
    """
    defaults = ROLE_DEFAULTS.get(role, _ALL_FALSE).copy()
    if not feature_access_json:
        return defaults
    try:
        overrides = json.loads(feature_access_json)
    except (json.JSONDecodeError, TypeError):
        return defaults
    if not isinstance(overrides, dict) or not overrides:
        # An EMPTY map is not a selection — it means "no per-user overrides", so
        # fall back to the role preset. (Unchecking every box in the admin UI
        # sends a full grid of false values, not {}, and that correctly grants
        # nothing via the fail-closed path below.)
        return defaults

    resolved = _ALL_FALSE.copy()
    for key, val in overrides.items():
        if key in resolved and isinstance(val, bool):
            resolved[key] = val
    return resolved

File: frontend/src/lib/features.ts
/**
 * Canonical feature list and role-based defaults.
 * Must stay in sync with backend/app/auth/features.py.
 */

export const FEATURE_CATEGORIES: Record<string, string[]> = {
  CRM: ['contacts', 'pipeline', 'tasks'],
  Sales: ['invoices', 'estimates', 'proposals'],
  Accounting: [
    'cashbook',
    'expenses',
    'smart_import',
    'email_scanner',
    'reports',
    'tax',
    'recurring',
  ],
  Communication: ['inbox', 'phone', 'sms'],
  Automation: ['workflows', 'forms'],
  Content: ['pages', 'docs', 'sheets', 'slides'],
  Storage: ['drive'],
  Meetings: ['calendar', 'meeting_rooms'],
  AI: ['obrain_chat', 'obrain_coach'],
  Admin: ['platform_admin', 'portal_admin'],
}

export const ALL_FEATURES = Object.values(FEATURE_CATEGORIES).flat()

const allTrue = Object.fromEntries(ALL_FEATURES.map(f => [f, true]))
const allFalse = Object.fromEntries(ALL_FEATURES.map(f => [f, false]))

export const ROLE_DEFAULTS: Record<string, Record<string, boolean>> = {
  admin: { ...allTrue },
  manager: { ...allTrue, platform_admin: false },
  team_member: { ...allTrue, platform_admin: false },
  accountant: {
    ...allFalse,
    cashbook: true,
    expenses: true,
    invoices: true,
    estimates: true,
    contacts: true,
    smart_import: true,
    email_scanner: true,
    reports: true,
    tax: true,
    recurring: true,
    drive: true,
  },
  client: { ...allFalse },
  // A viewer is a read-only collaborator: it owns nothing and sees only what is
  // shared with it. It needs the contacts module on, or it would be handed a
  // record it cannot open.
  viewer: { ...allFalse, contacts: true },
}

/**
 * Route prefix → the module it belongs to.
 *
 * Hiding a sidebar link does nothing to stop someone typing /cashbook into the
 * address bar, and every route in App.tsx was reachable that way. ProtectedRoute
 * resolves the LONGEST matching prefix, so new routes inherit a guard by default
 * rather than being silently unguarded.
 *
 * The backend is the real gate (require_feature). This is defence in depth, and
 * the difference between a bounced route and a page that renders then explodes
 * into 403s.
 */
export const ROUTE_FEATURES: Array<[string, string]> = [
  ['/contacts', 'contacts'],
  ['/tasks', 'tasks'],
  ['/invoices', 'invoices'],
  ['/estimates', 'estimates'],
  ['/proposals', 'proposals'],
  ['/cashbook', 'cashbook'],
  ['/reconcile', 'cashbook'],
  ['/expenses', 'expenses'],
  ['/income', 'expenses'],
  ['/budgets', 'expenses'],
  ['/reports', 'reports'],
  ['/recurring', 'recurring'],
  ['/smart-import', 'smart_import'],
  ['/email-scan', 'email_scanner'],
  ['/inbox', 'inbox'],
  ['/conversations', 'phone'],
  ['/communication', 'phone'],
  ['/drive', 'drive'],
  ['/documents', 'drive'],
  ['/meetings', 'meeting_rooms'],
  ['/calendar', 'calendar'],
  ['/bookings', 'calendar'],
  ['/availability', 'calendar'],
  ['/scheduling', 'calendar'],
  ['/pages', 'pages'],
  ['/forms', 'forms'],
  ['/workflows', 'workflows'],
  ['/docs', 'docs'],
  ['/obrain', 'obrain_chat'],
  ['/intelligence', 'obrain_coach'],
  ['/platform-admin', 'platform_admin'],
  ['/admin', 'platform_admin'],
]

/** The feature a path belongs to, or null if the path is always allowed. */
export function featureForPath(pathname: string): string | null {
  let best: [string, string] | null = null
  for (const entry of ROUTE_FEATURES) {
    if (pathname === entry[0] || pathname.startsWith(entry[0] + '/')) {
      if (!best || entry[0].length > best[0].length) best = entry
    }
  }
  return best ? best[1] : null
}

export const FEATURE_LABELS: Record<string, string> = {
  contacts: 'Contacts',
  pipeline: 'Pipeline',
  invoices: 'Invoices',
  proposals: 'Proposals',
  cashbook: 'Cashbook',
  smart_import: 'Smart Import',
  email_scanner: 'Email Scanner',
  reports: 'Reports',
  tax: 'Tax',
  recurring: 'Recurring',
  inbox: 'Inbox',
  phone: 'Phone',
  sms: 'SMS',
  pages: 'Pages',
  docs: 'Docs',
  sheets: 'Sheets',
  slides: 'Slides',
  drive: 'Drive',
  calendar: 'Calendar',
  meeting_rooms: 'Meeting Rooms',
  obrain_chat: 'O-Brain Chat',
  obrain_coach: 'O-Brain Coach',
  platform_admin: 'Platform Admin',
  portal_admin: 'Portal Admin',
}

export function getEffectiveFeatures(
  role: string,
  featureAccess: Record<string, boolean> | null | undefined,
): Record<string, boolean> {
  const defaults = { ...(ROLE_DEFAULTS[role] ?? allFalse) }
  if (featureAccess) {
    for (const [key, val] of Object.entries(featureAccess)) {
      if (key in defaults && typeof val === 'boolean') {
        defaults[key] = val
      }
    }
  }
  return defaults
}

/**
 * Does this user have the module?
 *
 * FAILS CLOSED. This used to `return true` when featureAccess was null, on a
 * "backward compat" rationale — which meant any user whose payload lacked the
 * field saw the entire nav. Combined with the backend never checking the flag at
 * all, a module being "off" meant nothing whatsoever.
 *
 * Callers must not evaluate this while the user is still loading, or it will
 * briefly deny everything. Gate on the auth store's loading state first —
 * ProtectedRoute does.
 */
export function hasFeature(
  featureAccess: Record<string, boolean> | null | undefined,
  key: string,
): boolean {
  if (!featureAccess) return false
  return featureAccess[key] === true
}
